AsyncDisplayKit/ASCollectionView.mm

```diff
@@ -313,6 +332,17 @@ - (void)setAsyncDelegate:(id<ASCollectionViewDelegate>)asyncDelegate
     super.delegate = (id<UICollectionViewDelegate>)_proxyDelegate;
     _asyncDelegateImplementsInsetSection = ([_asyncDelegate respondsToSelector:@selector(collectionView:layout:insetForSectionAtIndex:)] ? 1 : 0);
   }
+
+  [_flowLayoutInspector cacheSelectorsForCollectionView:self];
+}
+
+- (void)setCollectionViewLayout:(UICollectionViewLayout *)collectionViewLayout
+{
+  [super setCollectionViewLayout:collectionViewLayout];
+  if ([collectionViewLayout asdk_isFlowLayout]) {
+    _flowLayoutInspector = nil;
+    _layoutDelegate = [self flowLayoutInspector];
```

Comment:
I think what's not feeling right about this pattern is the fact that _flowLayoutInspector is set simply by calling the method...but you cannot call the method in all of the use cases that you need to access that object, as it will crash if you are not currently associated with a flow layout.

Please refactor -flowLayoutInspector to be something like -createInspectorForFlowLayout:()  (that's probably a bad name), explicitly passing in the flow layout and setting the _flowLayoutInspector property from the return value.  Then in this use case you will not need to nil out the property in advance.

Overall though I really feel like we need to consolidate or eliminate one of these properties.  It does not seem like there are many cases where _flowLayoutInspector is uniquely needed as opposed to calling _layoutDelegate, and having them be set to the same object makes that more unusual.